File: src/mapgen/starting_locations.cpp
//
// Created by Caelum van Ispelen on 9/2/21.
//

#include "starting_locations.h"

#include <queue>

namespace rip::mapgen {
    // Step 1: distribute players to specific continents.
    //
    // We prefer larger continents over smaller ones,
    // and we try to ensure each continent gets either zero or at least two players.
    std::vector<uint32_t>
    StartingLocationsGenerator::assignContinents(const std::vector<std::vector<glm::uvec2>> &continents, Rng &rng,
                                                 int numPlayers) {
        std::vector<uint32_t> playerContinentAssignments(numPlayers);

        for (int player = 0; player < numPlayers; player++) {
            std::vector<float> continentScores(continents.size());
            for (int continent = continents.size() - 1; continent >= 0; continent--) {
                // Count players on this continent.
                int numPlayersOnContinent = 0;
                for (int i = 0; i < player; i++) {
                    if (playerContinentAssignments[i] == continent) {
                        ++numPlayersOnContinent;
                    }
                }

                // Heuristic: each player consumes 100 tiles for a continent.
                // We assign a score based on the remaining tiles for this continent.
                continentScores[continent] = (continents[continent].size() - 100 * numPlayersOnContinent);

                // If the continent is very small, we should never assign it.
                if (continents[continent].size() < 30) {
                    continentScores[continent] = -1000000;
                }
            }

            // Select continent with the highest score.
            std::optional<int> bestContinent;
            for (int c = 0; c < continentScores.size(); c++) {
                if (!bestContinent.has_value()
                    || continentScores[c] > continentScores[*bestContinent]) {
                    bestContinent = c;
                }
            }

            playerContinentAssignments[player] = *bestContinent;
        }

        // Shuffle player continent assignments so player 0 doesn't
        // always get the biggest one.
        rng.shuffle(playerContinentAssignments.begin(), playerContinentAssignments.end());

        return playerContinentAssignments;
    }

    float StartingLocationsGenerator::scoreTile(const Grid<Tile> &tileGrid, glm::uvec2 pos,
                                                const std::vector<glm::uvec2> &otherStartingLocations) {
        const auto &tile = tileGrid.get(pos.x, pos.y);

        float score = 0;

        const auto terrain = tile.getTerrain();
        if (terrain == Terrain::Grassland) {
            score += 10;
        } else if (terrain == Terrain::Desert) {
            score -= 50;
        }

        if (tile.isHilled()) {
            score += 5;
        }

        for (const auto otherPos : otherStartingLocations) {
            float d = dist(pos, otherPos);

            score -= 100 / d;
        }

        // Big fat cross scores
        for (const auto bfcPos : getBigFatCross(pos)) {
            if (bfcPos.x >= tileGrid.getWidth() || bfcPos.y >= tileGrid.getHeight()) {
                // Out of bounds
                score -= 3;
                continue;
            }

            const auto &bfcTile = tileGrid.get(bfcPos.x, bfcPos.y);

            const auto bfcTerrain = bfcTile.getTerrain();
            if (bfcTerrain == Terrain::Grassland || bfcTerrain == Terrain::Ocean) {
                score += 2;
            } else if (bfcTerrain == Terrain::Desert) {
                score -= 2;
            } else if (bfcTerrain == Terrain::Plains) {
                score += 1;
            }
        }

        return score;
    }

    std::vector<glm::uvec2>
    StartingLocationsGenerator::generateStartingLocations(const Grid<LandCell> &landGrid, const Grid<Tile> &tileGrid,
                                                          Rng &rng, int numPlayers) {
        auto continents = landGrid.withAssignedIDs().groupToPositions;
        // Sort continents by size
        std::sort(continents.begin(), continents.end(), [](const std::vector<glm::uvec2> &a, const std::vector<glm::uvec2> &b) {
            return a.size() < b.size();
        });
        // Only respect land continents, not ocean ones
        for (int c = continents.size() - 1; c >= 0; c--) {
            auto pos = continents[c].at(0);
            if (landGrid.get(pos.x, pos.y) == LandCell::Ocean) {
                continents.erase(continents.begin() + c);
            }
        }

        // Get each player's continent.
        const auto continentAssignments = assignContinents(continents, rng, numPlayers);

        // Step 2: Assign each player a starting location on the continent
        // they were assigned.
        //
        // For each player, we give each tile in their continent a score
        // depending on its food yield, the number of adjacent deserts,
        // the distance to other players' starting locations,
        // and its coastal status.
        std::vector<glm::uvec2> startingLocations;

        struct TileWithScore {
            glm::uvec2 pos;
            float score;

            bool operator<(const TileWithScore &other) const {
                return score < other.score;
            }
        };

        for (int player = 0; player < numPlayers; player++) {
            std::priority_queue<TileWithScore> tileScores;
            for (const auto pos : continents[continentAssignments[player]]) {
                float tileScore = scoreTile(tileGrid, pos, startingLocations);

                tileScores.emplace(TileWithScore {
                    .pos = pos,
                    .score = tileScore,
                });
            }

            const auto posWithBestScore = tileScores.top().pos;
            startingLocations.push_back(posWithBestScore);
        }

        return startingLocations;
    }
}

File: crates/server/src/mapgen/starting_locations.rs
//! Generates a starting location for each player.
//!
//! This happens in two steps:
//! 1. Each player is assigned to a continent. We try to ensure
//! players are evenly distributede in terms of available land.
//! 2. Each player is assigned a starting location on their respective
//! continent. To do so, we assign a score to each land tile and use
//! the tile with the highest score.

use float_ord::FloatOrd;
use glam::{uvec2, UVec2};
use riposte_common::{Grid, Terrain};

use crate::game::Tile;

#[derive(Debug)]
struct Continent {
    tiles: Vec<UVec2>,
}

/// Locates each continent on a map. A continent
/// is a set of contiguous land tiles.
fn locate_continents(tiles: &Grid<Tile>) -> Vec<Continent> {
    let mut visited = Grid::<bool>::new(false, tiles.width(), tiles.height());

    let mut stack = Vec::new();

    let mut continents = Vec::new();

    for x in 0..tiles.width() {
        for y in 0..tiles.height() {
            let pos = uvec2(x, y);

            if *visited.get(pos).unwrap() {
                continue;
            }

            if tiles.get(pos).unwrap().terrain() == Terrain::Ocean {
                continue;
            }

            stack.clear();
            stack.push(pos);
            visited.set(pos, true).unwrap();

            let mut continent_tiles = Vec::new();

            while let Some(pos) = stack.pop() {
                continent_tiles.push(pos);

                for apos in tiles.adjacent(pos) {
                    if !*visited.get(apos).unwrap()
                        && tiles.get(apos).unwrap().terrain() != Terrain::Ocean
                    {
                        visited.set(apos, true).unwrap();
                        stack.push(apos);
                    }
                }
            }
            continents.push(Continent {
                tiles: continent_tiles,
            });
        }
    }

    continents
}

/// Distributes `n` players across the given list of continents.
///
/// Returned values are indices into `continents`.
fn assign_player_continents(continents: &[Continent], num_players: usize) -> Vec<usize> {
    let mut players_per_continent = vec![0; continents.len()];
    (0..num_players)
        .map(|_| {
            let continent_scores: Vec<_> = continents
                .iter()
                .enumerate()
                .map(|(i, continent)| {
                    let existing_players = players_per_continent[i];
                    // Heuristic: each player takes up 100 tiles on a continent.
                    let mut score = (continent.tiles.len() - existing_players * 100) as i32;

                    // If the continent is very small, players should avoid it.
                    if continent.tiles.len() < 30 {
                        score -= 100000;
                    }

                    score
                })
                .collect();

            let continent = continent_scores
                .iter()
                .enumerate()
                .max_by_key(|(_, score)| **score)
                .expect("zero continents")
                .0;
            players_per_continent[continent] += 1;
            continent
        })
        .collect()
}

/// Computes the score of a tile. Higher = better starting location.
fn score_tile(pos: UVec2, tiles: &Grid<Tile>, existing_starting_locations: &[UVec2]) -> f64 {
    let mut score = 0.;

    let tile = tiles.get(pos).unwrap();
    let terrain = tile.terrain();

    if tile.is_hilled() {
        score += 3.;
    }

    if terrain == Terrain::Grassland {
        score += 6.;
    } else if terrain == Terrain::Desert {
        score -= 10.;
    }

    if tile.has_fresh_water() {
        score += 60.;
    }

    // Prefer farther away from other starting locations
    for &other_pos in existing_starting_locations {
        score -= 100. / other_pos.as_f64().distance(pos.as_f64());
    }

    // Score each tile in the big fat cross of this tile
    for bfc_pos in tiles.big_fat_cross(pos) {
        let bfc_tile = tiles.get(bfc_pos).unwrap();
        let dscore = match bfc_tile.terrain() {
            Terrain::Ocean => 0.75,
            Terrain::Desert => 0.,
            Terrain::Plains => 0.75,
            Terrain::Grassland => 1.5,
            Terrain::Tundra => 0.25,
            Terrain::Mountains => -0.5,
        };
        score += dscore;
    }

    score
}

/// Generates players' starting locations.
pub fn generate_starting_locations(tiles: &Grid<Tile>, num_players: usize) -> Vec<UVec2> {
    let continents = locate_continents(tiles);
    let player_continents = assign_player_continents(&continents, num_players);

    let mut starting_locations = Vec::new();
    player_continents
        .iter()
        .map(|i| &continents[*i])
        .map(|continent| {
            let best_tile = continent
                .tiles
                .iter()
                .max_by_key(|&&pos| FloatOrd(score_tile(pos, tiles, &starting_locations)))
                .expect("continent is empty");

            starting_locations.push(*best_tile);

            *best_tile
        })
        .collect()
}
